File: webapp/codigos-fonte/Golang/lista.go
package main

import "fmt"

type No struct {
	Valor   int
	Proximo *No
}

type Lista struct {
	Topo    *No
	Tamanho int
}

func (l *Lista) inserir(valor, indice int) {
	if(indice > l.Tamanho){
		fmt.Printf("Não foi possível inserir, pois a lista tem %d elementos\n", l.Tamanho)
		return
	}
	novoValor := &No{
		Valor: valor,
	}

	var anterior *No = nil
	atual := l.Topo
	for i := 0; i < indice; i++ {
		anterior = atual
		atual = atual.Proximo
	}
	if anterior == nil {
		novoValor.Proximo = atual
		l.Topo = novoValor
	} else{
		novoValor.Proximo = atual
		anterior.Proximo = novoValor
	}
	l.Tamanho++
}

func (l *Lista) remover(indice int) {
	if l.Topo == nil {
		fmt.Printf("Erro ao tentar retirar um elemento de uma lista vazia\n")
		return
	}
	if l.Tamanho < indice {
		fmt.Printf("Erro ao tentar retirar um elemento inexistente de uma lista\n")
		return
	}
	var anterior *No = nil
	atual := l.Topo
	for i := 0; i < indice; i++ {
		anterior = atual
		atual = atual.Proximo
	}

	if anterior == nil {
		l.Topo = l.Topo.Proximo
	} else{
		anterior.Proximo = atual.Proximo
	}
}

func (l *Lista) printAll() {
	atual := l.Topo
	for atual != nil {
		fmt.Printf("%d ", atual.Valor)
		atual = atual.Proximo
	}
	fmt.Println()
}

func main() {
	l := &Lista{}
    l.inserir(5, 0)
    l.printAll()
    l.inserir(10, 0)
    l.printAll()
    l.inserir(15, 2)
    l.printAll()
    l.inserir(20, 1)
    l.printAll()
    l.remover(1)
    l.printAll()
    l.remover(0)
    l.printAll()
}

File: webapp/codigos-fonte/Kotlin/lista.kt
package lista

data class No (var valor:Int, var proximo: No? = null)

data class Lista (var topo:No? = null, var tamanho:Int = 0) {
    fun inserir(valor:Int, indice:Int) {
        if(indice > tamanho) {
            println(String.format("Não foi possível inserir, pois a lista tem %d elementos\n", this.tamanho))
            return
        }

        val novoValor = No(valor)

        var anterior:No? = null
        var atual:No? = this.topo

        for(i in 0 until indice) {
            anterior = atual
            atual = atual?.proximo
        }

        anterior?.let {
            novoValor.proximo = atual
            anterior.proximo = novoValor
        } ?: let {
            novoValor.proximo = atual
            this.topo = novoValor
        }
        this.tamanho++
    }

    fun remover(indice:Int) {
        this.topo?: let {
            println("Erro ao tentar retirar um elemento de uma lista vazia")
            return
        }
        if(this.tamanho < indice) {
            println("Erro ao tentar retirar um elemento inexistente de uma lista")
            return
        }

        var anterior:No? = null
        var atual:No? = this.topo

        for(i in 0 until indice) {
            anterior = atual
            atual = atual?.proximo
        }

        anterior?.let {
            anterior.proximo = atual?.proximo
        } ?: let {
            this.topo = this.topo?.proximo
        }
    }

    fun printAll() {
        var atual = this.topo
        while (atual != null) {
            print(String.format("%d ", atual.valor))
            atual = atual.proximo
        }
        println()
    }
}

fun main() {
    val l = Lista()
    l.inserir(5, 0)
    l.printAll()
    l.inserir(10, 0)
    l.printAll()
    l.inserir(15, 2)
    l.printAll()
    l.inserir(20, 1)
    l.printAll()
    l.remover(1)
    l.printAll()
    l.remover(0)
    l.printAll()
}